jetty-core/jetty-server/src/main/java/org/eclipse/jetty/server/handler/ConditionalHandler.java

```diff
@@ -0,0 +1,738 @@
+                    yield nextHandle.bindTo(nextNextHandler);
+                }
+                throw new IllegalStateException("No next Handler.Wrapper next");
+            }
+        };
+
+        // If we have method conditions, guard the doHandle method with checkMethod
+        if (!_methods.isEmpty())
+        {
+            MethodHandle checkMethod = lookup.findVirtual(ConditionalHandler.class, "checkMethod", methodType(boolean.class, Request.class)).bindTo(this);
+            _doHandle = MethodHandles.guardWithTest(
+                MethodHandles.dropArguments(checkMethod, 1, Response.class, Callback.class),
+                _doHandle,
+                doNotHandle);
+        }
+
+        // If we have path conditions, guard the doHandle method with checkPaths
+        if (!_paths.isEmpty())
+        {
+            MethodHandle checkPath = lookup.findVirtual(ConditionalHandler.class, "checkPath", methodType(boolean.class, Request.class)).bindTo(this);
+            _doHandle = MethodHandles.guardWithTest(
+                MethodHandles.dropArguments(checkPath, 1, Response.class, Callback.class),
+                _doHandle,
+                doNotHandle);
+        }
+
+        // If we have predicate conditions, guard the doHandle method with checkPredicates
+        if (!_predicates.isEmpty())
+        {
+            MethodHandle checkPredicates = lookup.findVirtual(ConditionalHandler.class, "checkPredicates", methodType(boolean.class, Request.class)).bindTo(this);
+            _doHandle = MethodHandles.guardWithTest(
+                MethodHandles.dropArguments(checkPredicates, 1, Response.class, Callback.class),
+                _doHandle,
+                doNotHandle);
+        }
+
+        super.doStart();
+    }
+
+    @Override
+    public boolean handle(Request request, Response response, Callback callback) throws Exception
+    {
+        try
+        {
+            // Invoke the _doHandle MethodHandle which will include checking any conditions in its guards.
+            return (boolean)_doHandle.invokeExact(request, response, callback);
+        }
+        catch (Exception e)
+        {
+            throw e;
+        }
+        catch (Throwable t)
+        {
+            Response.writeError(request, response, callback, t);
+            return true;
+        }
+    }
+
+    /**
+     * Handle a request that has met the conditions.
+     * The default implementation calls the {@link Handler#handle(Request, Response, Callback)} method
+     * on the {@link #getHandler() next Handler}.
+     * @param request The request to handle
+     * @param response The response to generate
+     * @param callback The callback for completion
+     * @return True if this handler will complete the callback
+     * @throws Exception If there is a problem handling
+     * @see Handler#handle(Request, Response, Callback)
+     */
+    protected boolean doHandle(Request request, Response response, Callback callback) throws Exception
```

Comment:
Rename to `didApply()` or `didMatch()` or `matched()`?

Reply:
I like `doHandle` because it follows the pattern we use elsewhere of a method like `start` delegating after some conditions to a `doStart` method.

Reply:
`doMatch()` along with `noMatch()` (instead of `doNotHandle()`, which seems a command)?

Reply:
@sbordet See comment above.  I don't think "match" is a concept that is immediately associated with the language of "meeting or not meeting a condition".     It makes sense to us because often a condition is a pattern that we do match, but some conditions may not be patterns.